pages/react-canvas-components.mdx

```diff
@@ -0,0 +1,256 @@
+  React.useEffect(() => {
+    const context2d = canvasRef.current.getContext('2d');
+    setRenderingContext(context2d);
+  }, []);
+
+  return (
+    <SharingContext.Provider value={renderingContext}>
+      <canvas ref={canvasRef} />
+      {/* the hexagons are passed through the `children` prop */}
+      {props.children}
+    </SharingContext.Provider>
+  );
+};
+```
+
+The `Hexagon` component needs to consume this React context to read its value – here with the `useContext` hook.
+
+```jsx hexagons-30dkf src/Hexagon.js
+const Hexagon = props => {
+  // we get the rendering context by comsuming the React context
+  const renderingContext = React.useContext(SharingContext);
+
+  if (renderingContext !== null) {
+    // hexagon drawing logic
+  }
+};
+```
+
+Now, both our `Canvas` and `Hexagon` components are ready. We can display several random hexagons:
+
+```jsx hexagons-30dkf src/App.js
+const App = () => (
+  <Canvas>
+    {getRandomHexagons().map(hexagon => (
+      <Hexagon {...hexagon} />
+    ))}
+  </Canvas>
+);
+```
+
+The last thing we need is to animate the hexagons so that they rotate.
+
+## 🎬 Animations
+
+As we saw, the canvas is like a sheet of paper: once we've drawn on it, we can't change it! But we can clear the canvas and re-draw on it. Animating on a canvas is like animating cartoons. It's a lots of draw, clean, draw, clean, repeat! To make a shape move, you need to split the movement into small steps, draw them one by one, while clearing the canvas between them. Those steps are called frames. Browsers come with an API `requestAnimationFrame` so that you can draw in each frame. 
+
+### 🖼 Creating a frame loop
+
+First things first, the canvas should be cleared at each frame. The easiest way to do this is to have an internal state counting the frames. This way, the component re-renders at each frame:
+
+```jsx animated-hexagons-v9j2q src/Canvas.js
+const [frameCount, setFrameCount] = React.useState(0);
+// this effect will increment frameCount by one at the next frame
+// and is called every time frameCount changes, creating a recursive loop
+// this makes the Canvas component re-render at every frame
+React.useEffect(() => {
+  const frameId = requestAnimationFrame(() => {
+    setFrameCount(frameCount + 1);
+  });
+  return () => {
+    cancelAnimationFrame(frameId);
+  };
+}, [frameCount, setFrameCount]);
+
+// here's the clearing at every render — at every frame.
+if (context !== null) {
+  context.clearRect(0, 0, actualWidth, actualHeight);
+}
+```
+
+But... the canvas is now white! It's because the hexagons are only rendered once – when `<RandomHexagons />` is first rendered. Yet, the canvas is cleared at every frame, erasing the hexagons on the second frame. Children components should also draw in the canvas at every frame. To force them to re-render, one solution is to share the `frameCount` from the canvas with the `Hexagon` component. It's achieved via a React context, like we did with `SharingContext`:
```

Comment:
- "It's because" => "This is because" 
- "This is because the hexagons are only rendered once - when `<RandomHexagons />` is first mounted. But, as the canvas is cleared on each frame, the hexagons are erased once the next render occurs"
- "Children components" => "Child components" 
- "Child components may also wish to draw in the canvas during a frame" 
- "It's achieved" => "This is achieved"

Reply:
I'm not sure about `may also wish` ? They _need_ to draw, it's not really an option.

Reply:
"should" was conditional too :) How about "Child components that draw in the canvas on every frame must be forced to re-render. One solution ... "

Reply:
True ! It's better to merge with the next sentence indeed. Will do